cmd/wfnconvert/options.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/facebookincubator/nvdtools/wfn"
+)
+
+type intFields []int
+
+// Set is part of flag.Value interface
+func (f *intFields) Set(s string) error {
+	for _, t := range strings.Split(s, ",") {
+		i64, err := strconv.ParseInt(t, 10, 64)
+		if err != nil {
+			return fmt.Errorf("failed to parse fields from %q: %v", s, err)
+		}
+		if i64 <= 0 {
+			return fmt.Errorf("illegal CSV index %d: %q", i64, s)
+		}
+		*f = append(*f, int(i64)-1)
+	}
+	return nil
+}
+
+// String is part of flag.Value interface
+func (f *intFields) String() string {
+	var sb strings.Builder
+	for i, n := range *f {
+		if i != 0 {
+			sb.WriteByte(',')
+		}
+		fmt.Fprintf(&sb, "%d", n+1)
+	}
+	return sb.String()
+}
+
+type strFields map[string]bool
+
+// Set is part of flag.Value interface
+func (f *strFields) Set(s string) error {
+	if *f == nil {
+		*f = make(strFields)
+	}
+	for _, s := range strings.Split(s, ",") {
+		(*f)[s] = true
+	}
+	return nil
+}
+
+// String is part of flag.Value interface
+func (f *strFields) String() string {
+	var sb strings.Builder
+	i := 0
+	for k := range *f {
+		if i != 0 {
+			sb.WriteByte(',')
+		}
+		sb.WriteString(k)
+		i++
+	}
+	return sb.String()
+}
+
+type options struct {
+	outBinding       string
+	attributes       strFields
+	invertAttributes bool
+	any2na, na2any   bool
+	csvFields        intFields
+	csvComma         string
+}
+
+func (o *options) addFlags() {
+	flag.StringVar(&o.outBinding, "b", "fstr", "ouput bindings, one of\n"+
+		"'uri'\te.g. cpe:/a:foo:bar:1.1)\n"+
+		"'fstr'\te.g. cpe:2.3:a:foo:bar:1.1:*:*:*:*:*:*:*)\n"+
+		"'str'\te.g. "+`wfn:[part="a",vendor="foo",product="bar", version="1\.1"])`+"\n")
+	flag.Var(&o.attributes, "attr", "comma-separated list of WFN attributes to manipulate; CPE 2.3 defines following attributes:\n"+
+		"part, vendor, product, version, update, edition, language, sw_edition, target_sw, target_hw, other\n"+
+		"additionally a special value 'all' is accepted")
+	flag.BoolVar(&o.invertAttributes, "v", false, "invert attributes: process attributes not matched by -v flag")
+	flag.BoolVar(&o.any2na, "any2na", false, "assign logical value N/A to all attributes with logical value ANY")
+	flag.BoolVar(&o.na2any, "na2any", false, "assign logical value ANY to all attributes with logical value N/A")
+	flag.Var(&o.csvFields, "csv", "comma-separated list of fields to read CPEs from, starting at 1;\n"+
+		"assumes input as CSV, fields not specified in the list are passed unchanged")
+	flag.StringVar(&o.csvComma, "d", ",", "if csv flag is set, sets the CSV delimiter, otherwise this flag is ignored")
+}
+
+func (o *options) validate() error {
+	if err := validateAttrNames(o.attributes); err != nil {
+		return err
+	}
+
+	if o.any2na && o.na2any {
+		return fmt.Errorf("-any2na and -na2any flags are mutually exclusive")
```

Comment:
why?
(NA, Any) should be converted to (Any, NA) if both specified.

Reply:
too complicated: i'll need to sequentially pass each attribute through all the transforms, instead of passing all of them through one, then all of them through the next one. since i don't see a practical application for that kind of stuff, i didn't bother.